### ISSUE TITLE
Automate creation of Table of Abbreviations and Acronyms

### ISSUE THREAD

**Issue by niazim3** [JacquesCarette/Drasil#698]
The table of abbreviations should eventually be automated like the table of symbols currently are. 

This would be ideal so that we won't have extra abbreviations or missing ones in the table that are(n't) present in the SRS documents. It would also avoid the need to make manual changes such as the ones made in #697.

**Comment by halonazhao** [comment]
Ok. After looking into this issue, I think it's better to do the reference improvement first.

 So far, the generated version has too many reference issues, like many IM reference are still hard-coded. 

Also in the generated version, technically there is even no appropriately generated  "A". The "A" is either hard-coded as "A" or shown up as "assumption" or "Assumption". The same thing for other abbreviations.

So at least, I should fix all the references issues first. Then I can probably start doing this based on "Reference", when although there will still be display issues like in manual version ("A1") but in drasil ("assump1".)
@szymczdm @niazim3

**Comment by JacquesCarette** [comment]
I think I agree that this issue is going to be easier to deal with once referencing has stabilized.

**Comment by halonazhao** [comment]
Some Analysis here:
1. Now the `Acronym` is constructed at type `CI`. 
`data CI = CI { _cid :: UID, _ni :: NP, _ab :: String}`
The abbreviation fills in `_ab :: String`.
2. Whether all `CommonIdea`s have abbreviations? 
According the field called `_ab` from CI, it looks like the short of abbreviation. But according to manual version, not all the abbreviations should be put in to table. I found `Matlab programming Language` defined but not showing in the table. Should this kind of abbreviation should be shown in the table too?
3. Taking SSP as example:
   

- `PS` shows in table but doesn't exsits in anywhere else in the document.
-  `DD, A, IM, GD, LC, R, T, UC`  show in the name of the references. But they are used as `String` in the names. So somehow we may want to make them to be constructed as type `CI` to replace the `String`.
-  `GS, SRS, SSA, Uncert` are in right type `CI`. So these should be able to be easily resolved.

Any thoughts here? @JacquesCarette @smiths Thanks!

**Comment by smiths** [comment]
@halonazhao, I don't think that Matlab should appear in the table of abbreviations and acronyms.  With respect to PS, it does appear in the document under Physical System descriptions.  Each physical system has the prefix PS (PS1, PS2, etc.)

**Comment by halonazhao** [comment]
Ok. Then I guess PS1 exists in other cases but not SSP, but that should be fine as long as it can be changed to use PS at type CI. I will fix this for SSP.

The problem here is that how do we decide which CI with Abbreviation should go into that table because obviously not all the CommonIdea with an abbreviation goes into that table. 

If we know specifically where those abbreviations should be collected from, the way to do collection would manually grab information from specific places.

If we don't, then we should figure out the differences of Abbreviations who go into the table and those who don't. And then design the technique to grab the right information.

These are two ways we design the collection method I guess.

**Comment by smiths** [comment]
@halonazhao, I inadvertently looked at SWHS, not SSP.  You are right that PS is not in SSP.  This is actually a problem with the case study.  SSP should mention the physical system being considered.  It isn't following the template by excluding this information.  @bmaclach and I are working on the SSP case study to fix it.  We'll be sure to add the PS information to it.  We should probably also add text to all examples that defines what PS stands for.  (Not just in the table of abbreviations, but in the main text.)

I wonder if abbreviations have been misused in the current Drasil code.  I would think that everything called an abbreviation would go into the list you are generating.  If something is called an abbreviation that shouldn't be in the list, then it isn't an abbreviation.  I think you Matlab example fit into this category.  (Although technically Matlab stands for Matrix Laboratory and could be considered an abbreviation, when we mention it we mean the software product Matlab, not the abbreviation.  However, if we explain the connection between the name Matlab and Matrix Laboratory, then I guess it could be in the table, but that seems awkward to me.)

**Comment by bmaclach** [comment]
@smiths I've already updated the manual case study with more information on the physical system (including an enumerated PS list) as part of my revision of SSP, just haven't made the corresponding update to Drasil yet.

Looks like @halonazhao has already made the change she needed to SSP to proceed with her work on the table of abbreviations, though it will probably get updated again when I get around to copying that change over to Drasil (I don't think that will interfere with your work, Yuzhi, since I will not be removing the PS acronym, but correct me if I'm wrong).

I also agree that we should add text explaining the acronym in the body of the SRS.

**Comment by smiths** [comment]
Great!  Thank you @bmaclach.  I forgot that you added the PS list.

Yes, you will likely update the Drasil version of SSP again when you proceed with your work.  @halonazhao is moving along quickly, so she will hopefully be done before you start.  That would mean that we wouldn't have to worry about conflicts between your work.

**Comment by JacquesCarette** [comment]
I am going to comment on this whole thread, starting from @halonazhao 's Analysis.  I want to confirm that indeed `CommonIdea`s always have an abbreviation -- otherwise they are a `NamedIdea`. (An `Idea` *may* have an abbreviation).

The next point is that we really have all sorts of different kinds of abbreviations. Some are standard nomenclature (such as for Units), some are domain-specific (in this case, for Slope Stability), some are Software Engineering (R for Requirement is probably such), while others still are from the document template (IM for Instance Model). There will most likely be other kinds.

So what we would need here (and don't but it is easy to fix) is something that is an instance of `ConceptDomain` without having a `Definition`. What all of these `CommonIdea`s are 'missing' some information, namely the context that they come from.  If these were in place, then in fact the Table of Abbreviation should really be a bunch of tables (or a single one, depending on the users' choice) for each of the given `ConceptDomain`. Which again matches well with the database-query model of assembling documents!

I guess the simplest fix would be to actually add `ConceptDomain` to `CommonIdea`. As it can be an empty list, that would port over very easily. Then we could incrementally add the proper information into our 'database' without breaking everything.

This will muck with @szymczdm 's tables in his current draft. But for the better!

**Comment by halonazhao** [comment]
So far, I've done collecting the explict`CI`s. But it turned out there are many implicit `CI`s being buried in `Sentences` using the `Idea` class. But `CI` is the only exit for `Idea` class to directly get a String. It's just that the `CI` would lost track when make it into a `Sentence`.

So the whole track is as below: 
https://github.com/JacquesCarette/Drasil/blob/541bfb80da69287710e3709680fa1a42cc3a0a47/code/drasil-example/Drasil/SWHS/Body.hs#L299

https://github.com/JacquesCarette/Drasil/blob/541bfb80da69287710e3709680fa1a42cc3a0a47/code/drasil-lang/Language/Drasil/Chunk/NamedIdea.hs#L11-L13

https://github.com/JacquesCarette/Drasil/blob/541bfb80da69287710e3709680fa1a42cc3a0a47/code/drasil-lang/Language/Drasil/Classes.hs#L51-L60

https://github.com/JacquesCarette/Drasil/blob/541bfb80da69287710e3709680fa1a42cc3a0a47/code/drasil-lang/Language/Drasil/Sentence.hs#L20-L23

So my thought is that we use a new constructor`Short UID` to take CI's UID to constructe a Sentence of a `CI`'s abbreviation. So that the every UID's of CI's would be collected properly. The only problem is that not all chunk with an UID would have an abbreviation. So I think we can use the same idea as the old `short` constructor, once there is no abbreviation for this chunk, we print the term as a sentence. 

Something like:
`Sentence = Short UID | ....`
`short :: Idea c => c -> Sentence`
`short c = maybe (phrase (c ^. term)) (Short $ c ^.uid) (getA c)`
What do you think? @JacquesCarette

**Comment by JacquesCarette** [comment]
There are multiple things going on here.

My first question is, why must the `CI` associated to `progName`, for the purposes of build a physical system description, be embedded in the `Sentence`? I guess this is part of trying to build the table of abbreviations automatically? I am going to guess that that is the case.

So I think the real problem is that `physSystDesc` takes a `Sentence` as its first argument. It should take an `Idea` instead. At least, that is my first thought.

I most definitely do not like the idea of adding `Short` to `Sentence`. I would prefer if `Sentence` had even fewer cases! However, adding a `Short` constructor to some other type **is** a good idea. Which goes back to the first problem: it is `physSystDesc` which needs 'fixed', as it boils things down to `Sentence` too soon. [And this is likely the case in other places too.]

**Comment by halonazhao** [comment]
Ok. So I think `physSystDescription` is not a good example here. This can be tackled as you said, taking the `Idea` directly.

I will give another example, 
https://github.com/JacquesCarette/Drasil/blob/30d8b5e9bf14396c80eea9e99c8ae966b3ab6c75/code/drasil-example/Drasil/SWHS/Body.hs#L134-L137
https://github.com/JacquesCarette/Drasil/blob/30d8b5e9bf14396c80eea9e99c8ae966b3ab6c75/code/drasil-example/Drasil/SWHS/Body.hs#L670-L683
https://github.com/JacquesCarette/Drasil/blob/30d8b5e9bf14396c80eea9e99c8ae966b3ab6c75/code/drasil-docLang/Drasil/DocumentLanguage.hs#L129

This one wouldn't be that easy to be addressed.

**Comment by JacquesCarette** [comment]
This is a better examples - but the solution is, I think, the same.  The `IntroProg` is meant to build the introduction to the program. It should take, as one of its inputs, the chunk associated to the program (as a concept).  In other words, in this case, I believe that the `swhs_pcm` argument should also be stored directly in the `IntroProg` constructor.  That would encode (part of) the **meaning** rather than just the text of that subsection.

The design idea here is to try not have have to dig into any `Sentence` to find **meaningful** information.  Usage information? Yes.

How many calls to `short` are there in the examples?

**Comment by halonazhao** [comment]
There are many `short` being used in each case and differently. 
![screen shot 2018-11-07 at 17 00 15](https://user-images.githubusercontent.com/28347043/48163892-ef31b800-e2ae-11e8-8a84-d844e41a8fd5.png)

These are introductions from SWHS and SSP. They do have some similarities but mostly they are customized and using the abbreviations wherever they need. No rules for that. So it is unlikely to have a template to take exactly how many `CI`s.

**Comment by halonazhao** [comment]
Unless `IntroProg` take [CI], but how to get a specific CI to use from the list would be the problem. Also, this way looks a little dumb.

**Comment by JacquesCarette** [comment]
Actually, in the two examples you give, the `short` is used in the same way -- once for SRS (which is what the document is), and once for the program that is the topic of the particular document.  You have to look at the overall structure, not the details of the text. So the use of SRS is implicit because of the document itself. So that's why `IntroProg` only needs to have one additional field, the `CI` that corresponds to the program.

The text itself would continue to be built in exactly the way it currently is.

**Comment by halonazhao** [comment]
I see. You mean there are only `srs` and `progName` in Intro sections. No more other abbreviations would be added in. Ok, I'll make the changes.

**Comment by JacquesCarette** [comment]
I think so. We should document that!  It's a Drasil assumption. [We need Drasil-in-Drasil, so we could indeed document that...  but that is not going to happen soon]

**Comment by halonazhao** [comment]
Aha, I just found one. `PCM` is another abbrev in IntroSection.  So the number of CI is different. But I can deal with others first. This is a complicated one.

**Comment by JacquesCarette** [comment]
The `CI` printed using `plural` (only) should be fine, because maybe their abbreviation is never used!  Even something has an abbreviation but we don't use it, it should not be collected.

But I am starting to see some value in having a more complex form of `Ch` in `Sentence`: we could have a new qualifier (to be named) that could be set to `Symbol` or `Term` or `Short` or `PluralTerm` (say) that would then get rendered that way.  The nice thing about this is that we wouldn't have to change our examples at all, only what the current smart constructors build [and the renderers, obviously].  The additional advantage would be that those smart constructors could have appropriate class constraints, which would ensure that the later lookup never fails.  [This isn't quite a design yet, definitely not a "please do this"; and yes, I do realize that this is a complex variant of something you proposed a couple of days ago!]

**Comment by halonazhao** [comment]
I see what you are saying here. It is a smart design. Why are we holding on this? Which part should be given more considerations?

**Comment by JacquesCarette** [comment]
For such changes in design, I always want to think about it at least overnight, as that's when I tend to find holes in the ideas. Larger changes, or deeper changes to core infrastructure, I often want to think about it for a couple of days.  Let's discuss at the meeting today.

The good news is that I have not found any holes in this idea (yet).

**Comment by halonazhao** [comment]
Summary of the process on this issue:
1. The structure of reorganizing the `Ch` to generate `Sentence` has been implemented.

2. The problem now is that there are some contents not matching `Stable`, and they are listed as following:
a. Constructor `the` problem causes some Upper Cases and Lower Cases not showing normally. A further discussion will be made in #1047.
b. DD, A, IM and so on should be collected from the name of the references. That's why they are not collected properly currently. So this part has to wait on `Reference Type` problem resolved.
c. Another problem is mention in the comment of 8467625b483915ffec853ee24b5b5b6585168d83.

3. The everything ChunkDB is currently roughly generated, definitely not complete but sufficient for each cases' uses. Is that ok just leave them as that for now?

**Comment by JacquesCarette** [comment]
1. I saw that. It looks great.
2. like the above,
  a. `the` was crafted quite carefully to deal with upper/lower cases, so this isn't entirely surprising.
  b. I don't quite understand this - more details please
  c. see comment there.
3. yes.

**Comment by halonazhao** [comment]
The `sentence` branch can match with `stable version`.

The problem `b` I talked about means that our reference name is like `DD:shearNorm` something like this. So the `DD` abbreviation should be collected from the name of the reference and only appear in the name. So this problem can be resolved when the `refType` problem is solved. But from the current view, if we do the `prefix` as a String. The abbreviation has to be reconsidered because it's no longer a CI collection problem.

**Comment by JacquesCarette** [comment]
I agree that the abbreviation problem is tied with the `refType` issue.  We should definitely not be collecting anything from names.

What we need to do is to collect abbreviations, however they occur (but doing it 'right', i.e. not dig into names). Indeed, it is not just a CI collection problem.

**Comment by halonazhao** [comment]
There are these two classes where we can potentially collect Abbreviations.

https://github.com/JacquesCarette/Drasil/blob/dc9a4aea2d83d172b9dbb9f87dda99283f1e3a7d/code/drasil-lang/Language/Drasil/Classes.hs#L56-L57
"instance Idea   DataDefinition where getA c = getA $ c ^. qd
instance Idea   InstanceModel where getA = getA . view rc
instance Idea   GenDefn where getA = getA . view relC
instance Idea   CI where getA = Just . view ab
instance Idea   ConstrainedChunk where getA = getA . view qd
instance Idea   ConstrConcept where getA = getA . view defq
instance Idea   TheoryModel where getA = getA . view con
instance Idea   UncertainChunk where getA (UCh n _) = getA n
instance Idea   UncertQ where getA (UQ q _) = getA q
instance Idea   NamedChunk where getA = \_ -> Nothing
instance Idea   IdeaDict where getA = view mabbr
instance Idea   DefinedQuantityDict where getA = getA . view con
instance Idea   QuantityDict where getA  qd = getA (qd ^. id')
instance Idea   UnitaryConceptDict where getA u = getA (u ^. unitary)
instance Idea   ConceptChunk where getA = getA . view idea
instance Idea   CommonConcept where getA = getA . view 
instance Idea   ConceptInstance where getA = getA . view (cc . idea)
instance Idea   RelationConcept where getA = getA . view conc
instance Idea   UnitaryChunk where getA uc = getA $ uc ^. quant
instance Idea   QDefinition where getA c = getA $ c ^. qua
instance Idea   UnitalChunk where getA (UC qc _) = getA qc
instance Idea   UnitDefn where getA c = getA (c ^. vc)"
https://github.com/JacquesCarette/Drasil/blob/dc9a4aea2d83d172b9dbb9f87dda99283f1e3a7d/code/drasil-lang/Language/Drasil/Classes.hs#L90-L92
instance CommonIdea         DataDefinition where abrv = abrv . view ci
instance CommonIdea         InstanceModel where abrv = abrv . view ci
instance CommonIdea         GenDefn where abrv = abrv . view ci
instance CommonIdea    CI where abrv = view ab
instance CommonIdea         TheoryModel where abrv = abrv . view ci
instance CommonIdea    AssumpChunk where abrv = abrv . view ci
instance CommonIdea    CommonConcept where abrv = abrv . view comm

I notice that `Idea` class implemented on `CI` too and the abbreviation of CI is collected from the same source as `CommonIdea` does. So we could ingore the `CommonIdea`'s effect on `CI`. But for `DataDefinition` and others, `Idea` gets abbreviation from 'IdeaDict' while 'CommonIdea' gets that from 'CI'.

**The first problem would be which source should our abbreviations collect from.**

**The second problem would be whether we need concept domain for `IdeaDict`.**
`IdeaDict` or other types constructed using `IdeaDict` could be made to `Sentence` using constructor `short`. `short` take a UID and print it's abbreviation. So if we have abbreviations from `short`, they need to be finally filtered by "conceptDomain" too. So should `IdeaDict` be attached with 'ConceptDomain'?

@JacquesCarette

**Comment by JacquesCarette** [comment]
First problem: this should not be an actual problem, as there should never be 2 conflicting dictionaries available. So whenever `getA` returns a `Just`, then assuming that chunk is used, its abbreviation should be picked up.  Note that it should be a mistake to implement `CommonIdea` but not `Idea`.

Second problem: I don't think so. `short` doesn't really print an abbreviation (if I remember well), it chooses between an abbreviation (if it exists) and a long form (otherwise).

**Comment by halonazhao** [comment]
The new approach was implemented in `TAandA` branch. The sentence set is collected from `Document` instead of `DocDesc` because more sentences were made during the rendering of `DocDesc` to `Document`. 

The result is correct. The only abbrevs left are `DD`, `IM`, `GD`, `IM`, `R`, `LC`.. These are supposed to be collected from the prefixes of the references.  @JacquesCarette

**Comment by JacquesCarette** [comment]
Sorry to take so long to get back to this.

I think collecting of `DD`, etc should still be done on `DocDesc`, i.e. by checking that the corresponding sections in the SRS document are non-empty.

**Comment by JacquesCarette** [comment]
I'm going to close this one in favour of #1661 .  Especially as I think this is, in fact, fixed.

**Comment by smiths** [comment]
I had a quick look at at Yuzhi's MEng report.  She only partially addressed abbreviations and acronyms.  She left completion of this task to future work.  I agree with closing this issue though, because #1661 supersedes it.
